app/javascript/components/util/DateInput.js

```diff
@@ -3,34 +3,85 @@ import PropTypes from 'prop-types';
 import { OverlayTrigger, Tooltip } from 'react-bootstrap';
 import DatePicker from 'react-datepicker';
 import MaskedInput from 'react-text-mask';
-
 import moment from 'moment';
+import _ from 'lodash';
 
 class DateInput extends React.Component {
   constructor(props) {
     super(props);
     this.datePickerRef = React.createRef();
-    this.handleRawChange = this.handleRawChange.bind(this);
-    this.handleDateChange = this.handleDateChange.bind(this);
-    this.clearDate = this.clearDate.bind(this);
+    this.state = {
+      currentDate: props.date || null,
+      lastValidDate: props.date || null,
+    };
   }
 
-  handleDateChange(date) {
-    this.props.onChange(date && moment(date).format('YYYY-MM-DD'));
+  /**
+   * Called only when date value is changed
+   */
+  handleDateChange = date => {
+    const momentDate = date && moment(date).format('YYYY-MM-DD');
+    this.props.onChange(momentDate);
     this.datePickerRef.current.setOpen(false);
-  }
+    if (this.dateIsValidAndNotEmpty(momentDate)) {
+      this.setState({ lastValidDate: momentDate });
```

Comment:
should `currentDate` be updated as well?

Reply:
We do not, that's where we start to see some issues because of the `onChange` happening before the `onBlur`.  We only want it to update with the `onRawChange` (as user is typing) and `onSelect` (when a date is selected in the datepicker)

Reply:
got it, would you mind adding some comments to explain when/why some of these functions are being called?

Reply:
updated here: https://github.com/SaraAlert/SaraAlert/pull/734/commits/3f3366ba02e8cc33155da0d952ce26786b27c687